Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp

```diff
@@ -234,18 +234,89 @@ void Jit64::dcbx(UGeckoInstruction inst)
+                         (js.op[1].inst.hex & 0xfc00'ffff) == 0x38000020 &&
+                         js.op[1].inst.RA_6 == inst.RB && js.op[1].inst.RD_2 == inst.RB &&
+                         js.op[2].inst.hex == 0x4200fff8;
+
   RCOpArg Ra = inst.RA ? gpr.Use(inst.RA, RCMode::Read) : RCOpArg::Imm32(0);
-  RCOpArg Rb = gpr.Use(inst.RB, RCMode::Read);
+  RCX64Reg Rb = gpr.Bind(inst.RB, make_loop ? RCMode::ReadWrite : RCMode::Read);
   RCX64Reg tmp = gpr.Scratch();
   RCX64Reg effective_address = gpr.Scratch();
   RegCache::Realize(Ra, Rb, tmp, effective_address);
 
+  RCX64Reg loop_counter;
+  if (make_loop)
+  {
+    // We'll execute somewhere between one single cacheline invalidation and however many are needed
+    // to reduce the downcount to zero, never exceeding the amount requested by the game.
+    // To stay consistent with the rest of the code we adjust the involved registers (CTR and Rb)
+    // by the amount of cache lines we invalidate minus one -- since we'll run the regular addi and
+    // bdnz afterwards! So if we invalidate a single cache line, we don't adjust the registers at
+    // all, if we invalidate 2 cachelines we adjust the registers by one step, and so on.
+
+    RCX64Reg& reg_cycle_count = tmp;
+    loop_counter = gpr.Scratch();
+    RegCache::Realize(loop_counter);
+
+    // This must be true in order for us to pick up the DIV results and not trash any data.
+    static_assert(RSCRATCH == Gen::EAX && RSCRATCH2 == Gen::EDX);
+
+    // Alright, now figure out how many loops we want to do.
+    const u8 cycle_count_per_loop =
+        js.op[0].opinfo->numCycles + js.op[1].opinfo->numCycles + js.op[2].opinfo->numCycles;
+
+    MOV(32, R(RSCRATCH), PPCSTATE(downcount));
+    MOV(32, R(reg_cycle_count), Imm32(cycle_count_per_loop));
+    MOV(32, R(loop_counter), PPCSTATE_CTR);
+
+    // This is both setting the adjusted loop count to 0 for the downcount <= 0 case and clearing
+    // the upper bits for the DIV instruction in the downcount > 0 case.
+    XOR(32, R(RSCRATCH2), R(RSCRATCH2));
+
+    TEST(32, R(RSCRATCH), R(RSCRATCH));                       // if (downcount <= 0)
+    FixupBranch downcount_is_zero_or_negative = J_CC(CC_LE);  // only do 1 invalidation; else:
+    DIV(32, R(reg_cycle_count));                              // RSCRATCH = downcount / cycle_count
+    LEA(32, RSCRATCH2, MDisp(loop_counter, -1));              // RSCRATCH2 = CTR - 1
+    // ^ Note that this CTR-1 implicitly handles the CTR == 0 case correctly.
+    CMP(32, R(RSCRATCH), R(RSCRATCH2));
+    CMOVcc(32, RSCRATCH2, R(RSCRATCH), CC_B);  // RSCRATCH2 = min(RSCRATCH, RSCRATCH2)
+
+    // RSCRATCH2 now holds the amount of loops to execute minus 1, which is the amount we need to
+    // adjust downcount, CTR, and Rb by to exit the loop construct with the right values in those
+    // registers.
+    SUB(32, PPCSTATE_CTR, R(RSCRATCH2));  // CTR -= RSCRATCH2
```

Comment:
Would it be faster to do this?

```suggestion
    SUB(32, loop_counter, R(RSCRATCH2));
    MOV(32, PPCSTATE_CTR, loop_counter);  // CTR -= RSCRATCH2
```

Reply:
I genuinely have no idea! Maybe? It technically avoids a read from memory, I suppose.

Reply:
Since I had an extra register available I also did this for the downcount memory write, along with moving some instructions into the `downcount > 0` branch. Results should still be identical, but please review again.